File: app/src/main/java/com/pr0gramm/app/services/RecentSearchesServices.java
package com.pr0gramm.app.services;

import android.content.SharedPreferences;

import com.google.common.collect.ImmutableList;
import com.google.common.reflect.TypeToken;
import com.google.gson.Gson;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.ArrayList;
import java.util.List;
import java.util.ListIterator;

import javax.inject.Inject;
import javax.inject.Singleton;

/**
 * Helps with recent searches
 */
@Singleton
public class RecentSearchesServices {
    private static final String KEY = "RecentSearchesServices.terms";
    private static final Logger logger = LoggerFactory.getLogger("RecentSearchesServices");

    private final List<String> searches = new ArrayList<>();
    private final SharedPreferences sharedPreferences;
    private final Gson gson;

    @Inject
    public RecentSearchesServices(SharedPreferences sharedPreferences, Gson gson) {
        this.sharedPreferences = sharedPreferences;
        this.gson = gson;

        restoreState();
    }

    public void storeTerm(String term) {
        synchronized (searches) {
            removeCaseInsensitive(term);
            searches.add(0, term);

            persistStateAsync();
        }
    }

    public ImmutableList<String> searches() {
        synchronized (searches) {
            return ImmutableList.copyOf(searches);
        }
    }

    public void clearHistory() {
        synchronized (searches) {
            searches.clear();
            persistStateAsync();
        }
    }

    /**
     * Removes all occurrences of the given term, independend of case.
     */
    private void removeCaseInsensitive(String term) {
        ListIterator<String> iter = searches.listIterator();
        while (iter.hasNext()) {
            if (iter.next().equalsIgnoreCase(term)) {
                iter.remove();
            }
        }
    }

    private void persistStateAsync() {
        try {
            // write down
            sharedPreferences.edit()
                    .putString(KEY, gson.toJson(searches, LIST_OF_STRINGS.getType()))
                    .apply();
        } catch (Exception ignored) {
            logger.warn("Could not presist recent searches");
        }
    }

    private void restoreState() {
        try {
            String serialized = sharedPreferences.getString(KEY, "[]");
            searches.addAll(gson.fromJson(serialized, LIST_OF_STRINGS.getType()));

        } catch (Exception error) {
            logger.warn("Could not deserialize recent searches", error);
        }
    }

    private static final TypeToken<List<String>> LIST_OF_STRINGS = new TypeToken<List<String>>() {
    };
}

File: app/src/main/java/com/pr0gramm/app/services/RecentSearchesServices.kt
package com.pr0gramm.app.services

import android.content.SharedPreferences
import androidx.core.content.edit
import com.pr0gramm.app.Logger
import com.pr0gramm.app.MoshiInstance
import com.pr0gramm.app.util.getStringOrNull
import java.util.*
import kotlin.reflect.javaType
import kotlin.reflect.typeOf


/**
 * Helps with recent searches
 */

class RecentSearchesServices(
        private val sharedPreferences: SharedPreferences) {

    private val logger = Logger("RecentSearchesServices")
    private val searches: MutableList<String> = ArrayList()

    init {
        restoreState()
    }

    fun storeTerm(term: String) {
        synchronized(searches) {
            removeCaseInsensitive(term)
            searches.add(0, term)

            persistStateAsync()
        }
    }

    fun searches(): List<String> {
        synchronized(searches) {
            return searches.toList()
        }
    }

    fun clearHistory() {
        synchronized(searches) {
            searches.clear()
            persistStateAsync()
        }
    }

    /**
     * Removes all occurrences of the given term, independend of case.
     */
    private fun removeCaseInsensitive(term: String) {
        searches.removeAll { it.equals(term, ignoreCase = true) }
    }

    private fun persistStateAsync() {
        try {
            // write searches as json
            val encoded = MoshiInstance.adapter<List<String>>(LIST_OF_STRINGS).toJson(searches)
            sharedPreferences.edit { putString(KEY, encoded) }
        } catch (ignored: Exception) {
            logger.warn { "Could not persist recent searches" }
        }

    }

    private fun restoreState() {
        try {
            val serialized = sharedPreferences.getStringOrNull(KEY) ?: "[]"
            searches.addAll(MoshiInstance.adapter<List<String>>(LIST_OF_STRINGS).fromJson(serialized)
                    ?: listOf())

        } catch (error: Exception) {
            logger.warn("Could not deserialize recent searches", error)
        }

    }

    companion object {
        private const val KEY = "RecentSearchesServices.terms"

        @OptIn(ExperimentalStdlibApi::class)
        @Suppress("PLATFORM_CLASS_MAPPED_TO_KOTLIN")
        private val LIST_OF_STRINGS = typeOf<java.util.List<String>>().javaType
    }
}
